cloud/pkg/upgradecontroller/controller/downstream.go

```diff
@@ -0,0 +1,400 @@
+	if isCompleted(upgrade) {
+		klog.Errorf("some upgrade is already upgrading or completed, don't send upgrade message again")
+		return
+	}
+
+	// get node list that need upgrading
+	var nodeName []string
+	if len(upgrade.Spec.NodeNames) != 0 {
+		for _, node := range upgrade.Spec.NodeNames {
+			nodeInfo, err := dc.kubeClient.CoreV1().Nodes().Get(context.Background(), node, metav1.GetOptions{})
+			if err != nil {
+				klog.Errorf("Failed to get node(%s) info: %v", node, err)
+				continue
+			}
+
+			if needUpgrade(nodeInfo, upgrade.Spec.Version) {
+				nodeName = append(nodeName, nodeInfo.Name)
+			}
+		}
+	} else if upgrade.Spec.LabelSelector != nil {
+		selector, err := metav1.LabelSelectorAsSelector(upgrade.Spec.LabelSelector)
+		if err != nil {
+			klog.Errorf("LabelSelector(%s) is not valid: %v", upgrade.Spec.LabelSelector, err)
+			return
+		}
+		nodes, err := dc.kubeClient.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{LabelSelector: selector.String()})
+		if err != nil {
+			klog.Errorf("Failed to get nodes with label %s: %v", selector.String(), err)
+			return
+		}
+
+		for _, node := range nodes.Items {
+			if needUpgrade(&node, upgrade.Spec.Version) {
+				nodeName = append(nodeName, node.Name)
+			}
+		}
+	} else {
+		klog.Errorf("Upgrade prohibit, no valid nodes are specified: no nodeNames and labelSelector are specified.")
+		return
+	}
+
+	// deduplicate: remove duplicate nodes to avoid repeating upgrade to the same node
+	nodeName = RemoveDuplicateElement(nodeName)
+
+	klog.Infof("Filtered finished, the below nodes are to upgrade\n%v\n", nodeName)
+
+	for _, node := range nodeName {
+		// send upgrade msg to every edge node
+		msg := model.NewMessage("")
+
+		resource := buildUpgradeResource(upgrade.Name, node)
+
+		// get HistoryID
+		var newHistoryID int
+		for index := range upgrade.Status.UpgradeStatus {
+			if upgrade.Status.UpgradeStatus[index].NodeName == node {
+				// historyID is 0,1,2...
+				newHistoryID = len(upgrade.Status.UpgradeStatus[index].History)
+				break
+			}
+		}
+		upgradeReq := commontypes.UpgradeRequest{
+			UpgradeID:       upgrade.Name,
+			UpgradeTool:     upgrade.Spec.UpgradeTool,
+			UpgradeType:     upgrade.Spec.UpgradeType,
+			HistoryID:       newHistoryID,
+			Version:         upgrade.Spec.Version,
+			ImageRepository: upgrade.Spec.ImageRepository,
+		}
+
+		msg.BuildRouter(modules.UpgradeControllerModuleName, modules.UpgradeControllerModuleGroup, resource, "upgrade").
+			FillBody(upgradeReq)
+
+		err := dc.messageLayer.Send(*msg)
+		if err != nil {
+			klog.Errorf("Failed to send upgrade message %v due to error %v", msg.GetID(), err)
+			continue
+		}
+
+		// process time out: cloud did not receive upgrade feedback from edge
+		// send upgrade timeout response message to upstream
+		go dc.handleUpgradeTimeout(node, upgrade.Name, upgrade.Spec.Version, newHistoryID, upgrade.Spec.TimeoutSeconds)
```

Comment:
it means that cloudcore didn't receive the upgrade msg, so it may not store upgrade status in Upgrade CR. That's what we should think about more. maybe in the next version.

Reply:
all the comments are fixed, except for the above two. PTAL again  @wackxu

Reply:
https://github.com/kubeedge/kubeedge/issues/4189